### PR TITLE
Milestone: Enumerable methods

### PR DESCRIPTION
**Pull request by meronogbai**
- Added methods to Enumerable module
- They are custom variants of built-in methods

### CODE REVIEW

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Meron Bsrat** 
Nice job with your first submission :muscle:  :muscle: 
Your project has failed some tests, check out the list below:

- [x] #my_each returns an Enumerator if no block is given
`[1,2,3].my_each.class == Enumerator`
- [x] #my_each when a block is given when self is a range returns the range itself after calling the given block once for each element in self
`[1,2,3].my_each(&proc{|x| x>2}) == [1,2,3]`
- [x] #my_each_with_index returns an enumerator if no block is given
`[1,2,3].my_each_with_index.class == Enumerator`
- [x] #my_each_with_index when a block is given when self is a range returns the range itself after calling the given block once for each element in self
`(1..3).my_each_with_index(&proc{|x| x>2}) == (1..3)`
- [x] #my_select returns an enumerator if no block is given
`[1,2,3,4].my_select.class==Enumerator`
- [x] #my_all? returns true if the block never returns false or nil
`[1,2,3].all?(&proc{|x| x>x/5}) == [1,2,3].my_all?(&proc{|x| x>x/5})`
- [x] #my_all? returns false if the block returns false or nil
`[1,2,3].all?(&proc{|x| x%2==0}) == [1,2,3].my_all?(&proc{|x| x%2==0})`
- [x] #my_all? when no block or argument is given returns true when none of the collection members are false or nil
`[true, [false]].all? == [true, [false]].my_all?`
- [x] #my_all? when no block or argument is given returns false when one of the collection members are false or nil
`[true,[true],false].all? == [true,[true],false].my_all?`
- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
`[1,2,3].all?(Integer) == [1,2,3].my_all?(Integer)`
- [x] #my_all? when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
`[1,-2,3.4].all?(Numeric) == [1,-2,3.4].my_all?(Numeric)`
- [x]  #my_all? when a class is passed as an argument returns false if any of the collection is not a member of such class
`['word',1,2,3].all?(Integer) == ['word',1,2,3].my_all?(Integer)`
- [x] #my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
`['car', 'cat'].all?(/a/) == ['car', 'cat'].my_all?(/a/)`
- [x] #my_all? when a Regex is passed as an argument returns false if any of the collection does not match the Regex
`['car', 'cat'].all?(/t/) == ['car', 'cat'].my_all?(/t/)`
- [x] #my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
`[5,5,5].all?(5) == [5,5,5].my_all?(5)`
- [x] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
`[5,5,[5]].all?(5) == [5,5,[5]].my_all?(5)`
- [x] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
`[0,[]].all? == [0,[]].my_all?`
- [x] #my_any? when no block or argument is given returns false if at least one of the collection is not true
`[false, 0].any? == [false, 0].my_any?`
- [x] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
`[1.1,'',[]].any?(Numeric) == [1.1,'',[]].my_any?(Numeric)`
- [x] #my_any? when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
`[1,'',[]].any?(Numeric) == [1,'',[]].my_any?(Integer)`
- [x] #my_any? when a Regex is passed as an argument returns true if any of the collection matches the Regex
`['dog', 'cat'].any?(/d/) == ['dog', 'cat'].my_any?(/d/)`
- [x] #my_any? when a Regex is passed as an argument returns false if none of the collection matches the Regex
`['dog', 'cat'].any?(/z/) == ['dog', 'cat'].my_any?(/z/)`
- [x] #my_any? when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
`['dog','car'].any?('cat') == ['dog','car'].my_any?('cat')`
- [x] #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
`['cat', 'dog','car'].any?('cat') == ['cat', 'dog','car'].my_any?('cat')`
- [x] #my_none? when no block or argument is given returns true only if none of the collection members is true
`[false, nil,false].none? == [false, nil,false].my_none?`
- [x] #my_none? when no block or argument is given returns false only if one of the collection members is true
`[false, nil,[]].none? == [false, nil,[]].my_none?`
- [x] #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class 
`[true,[]].none?(String) == [true,[]].my_none?(String)`
- [x] #my_none? when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
`[true,[]].none?(Numeric) == [true,[]].my_none?(Numeric)`
- [x] #my_none? when a class is passed as an argument returns false if any of the collection is a member of such class
`['',[]].none?(String) == ['',[]].my_none?(String)`
- [x] #my_none? when a Regex is passed as an argument returns true if none of the collection matches the Regex
`['dog','cat'].none?(/x/) ==  ['dog','cat'].my_none?(/x/)`
- [x] #my_none? when a Regex is passed as an argument returns false if any of the collection matches the Regex
`['dog', 'cat'].none?(/d/) == ['dog', 'cat'].my_none?(/d/)`
- [x] #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
`['dog','car'].none?(5) == ['dog','car'].my_none?(5)`
- [x] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
`[5,'dog','car'].none?(5) == [5,'dog','car'].my_none?(5)`
- [x] #my_count counts the number of items in enum that are equal to item if an argument is given
`[1,2,3].count(1) == [1,2,3].my_count(1)`
- [x] #my_count counts the number of elements yielding a true value if a block is given
`[1,2,3].count(&proc{|num|num%2==0}) == [1,2,3].my_count(&proc{|num|num%2==0})`
- [x] #my_map returns a new array with the results of running a given block once for every element in enum.
`[1,2,3].map{|x| x==2} == [1,2,3].my_map{|x| x==2}`
- [x] #my_map returns a new array with the results of running a given block once for every element in enum.
`[1,2,3].map(&proc{|x|x*2}) == [1,2,3].my_map(&proc{|x|x*2})`
- [x] #my_map returns an Enumerator if no block is given
`[1,2,3].map.class==[1,2,3].my_map.class`
- [x] #my_map accepts a proc as an argument and returns a new array with the results of running the proc once for every element in enum.
`[1,2,3].map(&proc{|x|x%2}) == [1,2,3].my_map(&proc{|x|x%2}) `
- [x] #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
`[1,2,3].my_map(proc{|x|x%2}){|a|a*2} == [1,2,3].my_map(proc{|x|x%2})`
- [x] #my_map does not mutate the original array
```
array = [1,2,3]
array_clone = array.clone
array.my_map{|num| num*2}
array == array_clone
```
- [x] #my_inject when a block is given with an initail value combines all elements of enum by applying a binary operation, specified by a block
`(1..3).inject(4) { |prod, n| prod * n } == (1..3).my_inject(4) { |prod, n| prod * n }`
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method
`[1,2,3].inject(:+) == [1,2,3].my_inject(:+)`
- [x] #my_inject when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method:range
`(1..9).inject(:+) == (1..9).my_inject(:+)`
- [x] #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method
`[1,2,3].inject(4, :*) == [1,2,3].my_inject(4, :*)`
- [x] #my_inject when a symbol is specified with an initail value combines each element of the collection by applying the symbol as a named method::range
`(1..3).inject(4, :*) == (1..3).my_inject(4, :*)`
- [x] #multiply_els accepts an array as an argument and multiplies all the elements of the array together using #my_inject
`multiply_els([1,2,3]) == 6`

Please request another review once you are ready! :+1: 

Happy coding :relaxed:

Oscar Andrés Russi Porras

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]
### Status: Changes Required ♻️ 
****
### Hi @meronokbay ! 👋 
You have done a great job implementing the changes requested by the previous TSE! 👍 There are just a few issues that need to be resolved before you can move on. Please take a look at the following methods.⬇️ 
****

- `my_all?`
- `my_map`

****
### Please make the required changes and resubmit for a new code review once you are ready 🙏

Good Luck and Happy Coding 🍀 👩🏾‍💻

_P.S. If you have any questions about my review or need help, you can message me on slack (@ Brittany Blake)._

File: script.rb [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+        return false unless element
+      end
+      true
+    elsif arg.class != Regexp && arg.class != Class
+      elements.my_each do |element|
+        return false unless element == arg
+      end
+      true
+    end
+  end
```
Comment (BrittanyBlake) [inline script.rb]: - [x] When a class is passed as an argument, `my_all?` should return true if all of the members of the collection is a member of the class (Numeric):
```ruby
 [1, 2i, 3.14].my_all?(Numeric) === [1, 2i, 3.14].all?(Numeric) => false
```

File: script.rb [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    elements.to_a
+    elements.my_each_with_index do |element, i|
+      elements[i] = if arg
+                      arg.call(element)
+                    else
+                      yield element
+                    end
+    end
+    elements
+  end
```
Comment (BrittanyBlake) [inline script.rb]: - [x] When `my_map` is called on a range, it should return a new array with the results of running a given block once for every element in enum.

```ruby
block = proc { |num| num < 4 }
range = Range.new(5, 50) 
range.my_map(&block) 
```

_Yours is currently returning a NoMethodError: `undefined method '[]=' for 5..50:Range`_

**Review by BrittanyBlake — CHANGES_REQUESTED** [review]

**Review by voscarmv — APPROVED** [review]
# Status: Project Approved ✨ 😎👍 💯 ✔️
### You may merge this branch and move forward ⛙🛣

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Pull request merged**